### DISCUSSION TITLE
Upgrade to react 18 and adopt suspense

### DISCUSSION THREAD

**Discussion by jacobhq** [vercel/swr#2313] [Ideas]
### Discussion in https://github.com/vercel/swr/discussions/1897

<div type='discussions-op-text'>

<sup>Originally posted by **jacobhq** March 30, 2022</sup>
Now that react 18 has been released, we will need to complete some tasks:
<br />
- [ ] Update callouts in `vercel/swr-site` that say suspense is in beta
- [ ] Increase prominence of suspense related documentation
- [ ] Discuss whether suspense should be used by default in swr
</div>

**Comment by shuding** [comment]
Thanks for creating this! 

Still there're a few things about clarification when using Suspense in SWR (or any other data fetching library). I personally still don't recommend using it in production:
- When using Suspense with SSR, it will likely to cause hydration mismatch issues: #1832
- Waterfall and conditional requests with Suspense isn't solved: #168

Last year I shared this write-up to the React working group: [**The Journey of SWR and Suspense**](https://gist.github.com/shuding/6ef6a85c4c8ee57d9926e705adef88e3), to give feedback of using Suspense for data fetching. As you can see that today, there are still a lot of things missing. And it's very possible that the API of using Suspense in SWR will change quite a bit. It sounds good to change it to "beta" but a lot of research is still needed (I know that there're other libs trying to solve these problems too).

Also from the [official blog post of React 18](https://reactjs.org/blog/2022/03/29/react-v18.html#suspense-in-data-frameworks):

> Ad hoc data fetching with Suspense is technically possible, but still not recommended as a general strategy.

**Comment by stefee** [comment]
The comment about ‘ad hoc’ I interpret as meaning ‘don’t try to do fetching without a data fetching library’ not ‘data fetching libraries should not adopt suspense’?

**Comment by shuding** [comment]
Sorry I didn't mean that data fetching libraries should not use Suspense. What I'm thinking is Suspense based data fetching is almost impossible to implement correctly in the library level as of today, it should be done to the framework level.

**Comment by stefee** [comment]
Ah I see. Thanks for clarifying this.

**Comment by jacobhq** [comment]
Thanks for that, so are we waiting for the mentioned issues to be solved, and for suspense to become more stable?

**Comment by pkellner** [comment]
> Thanks for that, so are we waiting for the mentioned issues to be solved, and for suspense to become more stable?

I would love to know the answer to this question.  I posted an issue here yesterday (https://github.com/vercel/swr/issues/1906) that makes me feel that SWR is completely broken with NextJS in React 18.  I'm having trouble believing no one is talking about this but just a few of us over here.  React 18 is a released product from a team the prides itself on not releasing stuff until it's ready.  What's going on?

BTW, i did test my little app with CRA and it worked with no error, but since it's a server rendering issue of sorts, I guess that's not surprising.

**Comment by stefee** [comment]
React 18 is released but a major version change. It will take time for the ecosystem of libraries and frameworks to be ready, so we have to be patient. I’m sure @shuding can add some more clarity on what level of stability we can expect at the moment.

**Comment by shuding** [comment]
👋 Hey! I answered @pkellner's question [here](https://github.com/vercel/swr/issues/1906#issuecomment-1094416833), as I said:

> I will keep this issue open until we add more clarification and guidance on Suspense and SSR to the documentation. Let me know if there's any suggestion on this!

As for level of stability for expectation (_of using React 18 + Suspense + SWR, probably Next.js too_), here're my thoughts:
- Suspense is still experimental in SWR, so there is a risk of using beta feature. We can't make it stable because:
  1. It can't cover all the SWR functionalities yet. Such as avoiding waterfall and dependent fetching.
  2. It can't cover data fetching use cases very well in SSR (that many frameworks do), see below.
- React 18 is a major version change as @stefee mentioned, and it does change the [behavior](https://github.com/reactjs/rfcs/blob/main/text/0213-suspense-in-react-18.md#summary) especially for Suspense. When upgrading from 17 to 18,
  1. Suspense will be executed during SSR, so make sure the SWR request works on the server side.
  2. Since Suspense will be executed during SSR, the rendered result will be hydrated too. If the requested and rendered data changes between SSR and hydration, you will get a hydration mismatch error. So make sure to not fetch dynamic data using Suspense + SSR.

**In short, if you are using a SSR framework, and using Suspense to fetch data from an API (rather than static resource), it's not safe.** 

But, if your application doesn't have the concerns above, e.g. you are using create-react-app, or the data is mostly static, then you should be fine.

Note that Next.js will always do SSR for static pages (https://nextjs.org/docs/advanced-features/automatic-static-optimization), that counts too. Other popular frameworks also do SSR. Create-react-app only does CSR.

And lastly, this is **not** a problem of Suspense in SWR. Every Suspense based data fetching library will have these problems.

**Comment by pkellner** [comment]
related https://github.com/reactjs/rfcs/pull/229
